cocotb/regression.py

```diff
@@ -229,63 +230,111 @@ def next_test(self):
             return None
         return self._queue.pop(0)
 
-    def _add_failure(self, result):
-        self.xunit.add_failure(stdout=repr(str(result)),
-                               stderr="\n".join(self._running_test.error_messages),
-                               message="Test failed with random_seed={}".format(self._seed))
-        self.failures += 1
-
     def handle_result(self, test):
         """Handle a test completing.
 
-        Dump result to XML and schedule the next test (if any).
+        Dump result to XML and schedule the next test (if any). Entered by the scheduler.
 
         Args:
             test: The test that completed
         """
         assert test is self._running_test
 
-        real_time   = time.time() - test.start_time
+        real_time = time.time() - test.start_time
         sim_time_ns = get_sim_time('ns') - test.start_sim_time
-        ratio_time  = self._safe_divide(sim_time_ns, real_time)
-        
+        ratio_time = self._safe_divide(sim_time_ns, real_time)
+
         self.xunit.add_testcase(name=test.funcname,
                                 classname=test.module,
                                 time=repr(real_time),
                                 sim_time_ns=repr(sim_time_ns),
                                 ratio_time=repr(ratio_time))
 
+        # score test
+        result_pass, sim_failed = self.score_test(test, test._outcome)
+
+        # Save results
+        self._store_test_result(test.__module__, test.__name__, result_pass, sim_time_ns, real_time, ratio_time)
+        if not result_pass:
+            self.xunit.add_failure()
+            self.failures += 1
+
+        # Fail if required
+        if sim_failed:
+            self.tear_down()
+
+        self.execute()
+
+    def _init_test(self, test_func):
+        test_init_outcome = cocotb.outcomes.capture(test_func, self._dut)
+
+        if isinstance(test_init_outcome, cocotb.outcomes.Error):
+            self.log.error("Failed to initialize test %s" % test_func.name, exc_info=True)
+            self.xunit.add_testcase(name=test_func.name,
+                                    classname=test_func.__module__,
+                                    time="0.0",
+                                    sim_time_ns="0.0",
+                                    ratio_time="0.0")
+            result_pass, sim_failed = self.score_test(test_func, test_init_outcome.without_frames(['capture']))
+            # Save results
+            self._store_test_result(test_func.__module__, test_func.__name__, result_pass, 0.0, 0.0, 0.0)
+            if not result_pass:
+                self.xunit.add_failure()
+                self.failures += 1
+            # Fail if required
+            if sim_failed:
+                self.tear_down()
```

Comment:
This won't actually stop subsequent tests being run, will it? I think that's the intent of `tear_down`.

Reply:
`sim_failed` is only true if `SimFailure` was the result and `tear_down` causes cocotb to stop. I'm not sure how possible that is for initialization, but it's a potential result so it's covered.

Reply:
Not sure I got my point across very well there.

As written, if test1 causes the simulator to die (which as you correctly remark is unlikely in the first place), then the regression manager will still try to start all the remaining tests - stopping the simulator isn't going to immediately stop the regression manager setup code.

Perhaps since this is so unlikely anyway, you should just raise `RuntimeError` here after calling `tear_down`, since if the simulator really did die in setup something has gone very wrong.

Reply:
`tear_down` calls `simulator.stop_simulator` which calls `gpi_sim_end`, so the regression manager code stops there immediately. `tear_down` is basically an exit function, it can't be run more than once.

Reply:
I added an exception here anyway with a message indicating that it should not be possible.